gtsfm/frontend/correspondence_generator/keypoint_aggregator/keypoint_aggregator_dedup.py

```diff
@@ -0,0 +1,110 @@
+"""Keypoint aggregator that de-duplicates keypoint detections, aggregating from image pairs to single images.
+
+Authors: John Lambert
+"""
+
+from typing import Dict, List, Optional, Tuple
+
+import numpy as np
+
+import gtsfm.utils.logger as logger_utils
+from gtsfm.common.keypoints import Keypoints
+from gtsfm.frontend.correspondence_generator.keypoint_aggregator.keypoint_aggregator_base import KeypointAggregatorBase
+
+logger = logger_utils.get_logger()
+
+
+class KeypointAggregatorDedup(KeypointAggregatorBase):
+    """Keypoint aggregator with de-duplication."""
+
+    def __init__(self) -> None:
+        """Initialize global variables"""
+        self.duplicates_found = 0
+
+    def append_unique_keypoints(
+        self, i: int, keypoints: Keypoints, per_image_kpt_coordinates: Dict[Tuple[int, int], np.ndarray]
+    ) -> Tuple[Dict[Tuple[int, int], np.ndarray], np.ndarray]:
+        """Identify unique keypoints, and append them to running list of global keypoints per image.
+
+        If duplicate keypoints are found, the index of the previously existing keypoint is recorded.
+
+        Args:
+           i: image frame index.
+           keypoints: keypoints detected in single image, from direct image pair feature matching.
+           per_image_kpt_coordinates: running list of global keypoints, per image.
+
+        Returns:
+            per_image_kpt_coordinates: running list of global keypoints, per image.
+            i_indices: single column of putative correspondence indices, for pair. These represent indices
+                into the global table of keypoints per image.
+        """
+        N_to_add = keypoints.coordinates.shape[0]
+        i_indices = np.zeros(N_to_add)
+        i_count = per_image_kpt_coordinates[i].shape[0]
+        unique_keypoints_i_coordinates = []
+
+        for k, uv in enumerate(keypoints.coordinates):
+            diff_norms = np.linalg.norm(per_image_kpt_coordinates[i] - uv, axis=1)
+            is_identical = np.any(diff_norms == 0)
```

Comment:
Should we make it this strict? Why not have some threshold?

Reply:
It's something that could be tweaked or modified for sure. I'll leave a todo.